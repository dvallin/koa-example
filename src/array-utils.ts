function isDefined<T>(t: T | undefined): boolean {
  return t !== undefined
}

export function pluck<T, K extends keyof T>(values: T[], key: K): T[K][] {
  return values.map(v => v[key]).filter(isDefined)
}
