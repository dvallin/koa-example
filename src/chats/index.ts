import * as IO from '../io'
import { Module } from '..'
import { buildChatSocketHandler } from './chat-socket-handler'
import { ChatServiceImpl, ChatService } from './chat-service'
import { ChatMessageRepositoryImpl, ChatMessageRepository, migrations } from './chat-message-repository'

export interface Components {
  messageRepository: ChatMessageRepository
  service: ChatService
}

export interface Message {
  room: string
  date: Date
  message: string
}

export function production(io: IO.Components): Module<Components> {
  const messageRepository = new ChatMessageRepositoryImpl(io.postgres)
  const service = new ChatServiceImpl(messageRepository)
  const chatSocketHandler = buildChatSocketHandler(service)
  return {
    components: { messageRepository, service },
    exports: { socketHandlers: [chatSocketHandler], migrations: migrations },
  }
}
