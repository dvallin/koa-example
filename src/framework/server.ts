// koa request id is nearly impossible to require with import
// eslint-disable-next-line @typescript-eslint/no-var-requires
const requestId = require('koa-requestid')

import * as Koa from 'koa'
import * as SocketIO from 'socket.io'
import { errorHandler } from './instrumentation/error-handler'
import { LoggerProvider } from './logger'
import { Server } from 'http'
import { SocketHandler, KoaHandler } from './service/handlers'

export function startApp(app: Koa, port: number | undefined = undefined): Server {
  return app.listen(port || process.env.SERVER_PORT)
}

export function build(middlewares: KoaHandler[], loggerProvider: LoggerProvider): Koa {
  const app = new Koa()

  app.use(requestId())
  app.use(errorHandler(loggerProvider('MainErrorHandler')).handler)
  middlewares.forEach(m => app.use(m.handler))

  return app
}

export function registerSockets(server: Server, socketHandlers: SocketHandler[]): void {
  const io = SocketIO(server)
  socketHandlers.forEach(handler => io.on('connect', socket => handler.handler(io, socket)))
}
