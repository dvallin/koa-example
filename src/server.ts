import * as Koa from 'koa'
import * as SocketIO from 'socket.io'
import { Server } from 'http'

// koa request id is nearly impossible to require with import
// eslint-disable-next-line @typescript-eslint/no-var-requires
const requestId = require('koa-requestid')

import { Mode, SocketHandler } from './'
import { LoggerProvider, Logger } from './io/logger'

function errorHandler(logger: Logger) {
  return async (ctx: Koa.Context, next: () => Promise<{}>) => {
    try {
      await next()
    } catch (err) {
      ctx.status = err.status || 500
      ctx.body = err.details || err.msg || 'Internal Server Error'
      if (ctx.status === 500) {
        logger.error(err)
      }
    }
  }
}

export function startApp(app: Koa, port: number | undefined = undefined): Server {
  return app.listen(port || process.env.SERVER_PORT)
}

export function registerSockets(server: Server, socketHandlers: SocketHandler[]): void {
  const io = SocketIO(server)
  socketHandlers.forEach(handler => io.on('connect', socket => handler(io, socket)))
}

export function build(middlewares: Koa.Middleware[], loggerProvider: LoggerProvider): Koa {
  const app = new Koa()

  app.use(requestId())
  app.use(errorHandler(loggerProvider('MainErrorHandler')))
  middlewares.forEach(m => app.use(m))

  return app
}

export async function startMode(mode: Mode, port: number | undefined = undefined): Promise<Server> {
  await mode.exports.startup()
  await mode.io.postgres.performTransaction(mode.exports.migrations)
  const server = await startApp(build(mode.exports.middlewares, mode.io.loggerProvider), port)
  server.on('close', mode.exports.shutdown)
  process.on('SIGTERM', () => {
    server.close(() => {
      process.exit(0)
    })
  })
  registerSockets(server, mode.exports.socketHandlers)
  return server
}
