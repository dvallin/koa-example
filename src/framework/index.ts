import { pluck } from './array-utils'
import { LoggerProvider } from './logger'

export type Hook = () => Promise<void>

export interface ModuleExports<T> {
  startup: Hook
  shutdown: Hook

  handlers: T[]

  ready: () => boolean
}

export interface NoneHandler {
  type: 'none'
  handler: () => never
}

export interface Module<T, H = NoneHandler> {
  components: T
  exports: Partial<ModuleExports<H>>
}

export type Mode<T, H> = ModeBase<H> & { components: T }

export interface ModeBase<H> {
  loggerProvider: LoggerProvider
  exports: ModuleExports<H>
}

async function callHooks(hooks: Hook[]): Promise<void> {
  await Promise.all(hooks.map(hook => hook()))
}

export function mergeExports<T>(...moduleExports: Partial<ModuleExports<T>>[]): ModuleExports<T> {
  const startup = (): Promise<void> => callHooks(pluck(moduleExports, 'startup'))
  const shutdown = (): Promise<void> => callHooks(pluck(moduleExports, 'shutdown'))
  const handlers = [].concat(...pluck(moduleExports, 'handlers'))
  const ready = () => pluck(moduleExports, 'ready').every(ready => ready())
  return { startup, shutdown, handlers, ready }
}
