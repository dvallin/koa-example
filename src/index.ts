import * as Users from './users'
import * as Chats from './chats'
import * as IO from './io'
import { pluck } from './array-utils'
import { QueryConfig } from 'pg'
import { Middleware } from 'koa'
import { Socket, Server } from 'socket.io'

export type Hook = () => Promise<void>
export type SocketHandler = (server: Server, socket: Socket) => void

export interface ModuleExports {
  startup: Hook
  shutdown: Hook

  middlewares: Middleware[]
  migrations: QueryConfig[]
  socketHandlers: SocketHandler[]
}

export interface Module<T> {
  components: T
  exports: Partial<ModuleExports>
}

export interface Mode {
  io: IO.Components
  users: Users.Components
  chats: Chats.Components

  exports: ModuleExports
}

async function callHooks(hooks: Hook[]): Promise<void> {
  await Promise.all(hooks.map(hook => hook()))
}

export function mergeExports(...moduleExports: Partial<ModuleExports>[]): ModuleExports {
  const startup = (): Promise<void> => callHooks(pluck(moduleExports, 'startup'))
  const shutdown = (): Promise<void> => callHooks(pluck(moduleExports, 'shutdown'))
  const migrations = [].concat(...pluck(moduleExports, 'migrations'))
  const middlewares = [].concat(...pluck(moduleExports, 'middlewares'))
  const socketHandlers = [].concat(...pluck(moduleExports, 'socketHandlers'))
  return { startup, shutdown, migrations, middlewares, socketHandlers }
}

export function production(): Mode {
  const io = IO.production()
  const users = Users.production(io.components)
  const chats = Chats.production(io.components)
  return {
    io: io.components,
    users: users.components,
    chats: chats.components,
    exports: mergeExports(io.exports, users.exports, chats.exports),
  }
}
