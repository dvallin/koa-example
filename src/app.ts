import { start } from './server'
import { production } from './mode'

export const server = start(production())
