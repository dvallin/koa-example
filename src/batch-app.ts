import { runMode } from './framework/batch/runner'
import { mergeExports } from './framework'
import { ConsoleLogger } from './framework/logger'
import { wrapBatchHandler } from './framework/batch/handlers'

runMode({
  loggerProvider: (name: string) => new ConsoleLogger(name),
  exports: mergeExports({
    handlers: [wrapBatchHandler(() => Promise.resolve())],
  }),
})
