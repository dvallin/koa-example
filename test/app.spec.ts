import { startMode } from '../src/server'
import { testing } from './test-mode'
import { MockPostgres } from './database/mock-database'
import { Server } from 'http'

import * as request from 'supertest'

describe('app', () => {
  let postgres: MockPostgres
  let server: Server
  beforeEach(async () => {
    const mode = testing()
    postgres = mode.database.postgres as MockPostgres
    server = await startMode(mode)
  })

  it('returns user name', async () => {
    postgres.backend.mockReturnValue([{ name: 'mock user name' }])

    const response = await request(server).get('/users')
    expect(response.status).toEqual(200)
    expect(response.text).toEqual('mock user name')
  })

  it('runs migrations', async () => {
    expect(postgres.migrations).toMatchSnapshot()
  })
})
