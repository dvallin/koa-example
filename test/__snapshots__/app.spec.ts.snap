// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app runs migrations 1`] = `
Array [
  Object {
    "config": Object {
      "name": "create-user-table",
      "text": "
      CREATE TABLE IF NOT EXISTS koa_users (
        email  VARCHAR(45) PRIMARY KEY,
        name   VARCHAR(45)
      );",
    },
  },
]
`;
